<a href="https://colab.research.google.com/github/llRodroll/ML/blob/main/TF/TF_Lab_2_Comp_Vis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Computer Vision

A NN to recognize cloth items from a 70,000 data poitns in the dataset Fashion MNIST. Each image falls in either of 10 categories.

## Libraries

In [1]:
import tensorflow as tf          
import matplotlib.pyplot as plt  # graphics

## Import Dataset

In [2]:
mnist = tf.keras.datasets.fashion_mnist
(training_images, training_labels), (test_images, test_labels) = mnist.load_data()

4423680/4422102 [==============================] - 0s 0us/step


## Data Visualization

The values representing the data are integers between 0 and 255.

The image is in a 28x28 grid of greyscale colors. Nevetherless, the images in the output do not appear that way.

In [ ]:
plt.imshow(training_images[0])
print(training_labels[0])
print(training_images[0])

To better manage the process iti is convinient to normalize the data between 0 and 1.

In [5]:
training_images  = training_images / 255.0
test_images = test_images / 255.0

## Model Design

For now the model have 3 layers:

In [7]:
model = tf.keras.models.Sequential([tf.keras.layers.Flatten(), 
                                    tf.keras.layers.Dense(128, activation=tf.nn.relu), 
                                    tf.keras.layers.Dense(10, activation=tf.nn.softmax)])

**Sequential**: defines the NN layers and its features.

**Faltten**: takes de 28x28 input (image) and turns it into a one-dimensional vector of $28\times28=784$ elements.

**Dense**: these are layers to procces the information. The first one has 128 nodes (neurons) and the second has 10 nodes (the number of label categories).

**Activation**: this tells the NN neurons what to do. In this case, we have two types. Relu is a function with 0 valuer for negative values and Softmax simply identfies the maximum value of a vector with the number 1 in that position and the rest is zero.

## Compile the model

As in previous labs, here we compile the model and choose some options.

**Adam** is an option to avoid saddle points in the optimization process.

**Loss** is the function to optimize. In this case we want to minimize our losses with the crossentropy.

**Metrics** is the metric that the training process will show. In this case is accuracy, i.e. the percentage of images correctly predicted.

In [8]:
model.compile(optimizer = tf.keras.optimizers.Adam(),
              loss = 'sparse_categorical_crossentropy',
              metrics=['accuracy'])

## Train the model

Remember that *epochs* is the number of iterations (tries) for the fit process. In this case only 5, a very low number.

Just to keep in mind: The *batch* size is a number of samples processed before the model is updated. The number of *epochs* is the number of complete passes through the training dataset. In this example we do not use batch explicity.

In [9]:
model.fit(training_images, training_labels, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.5005 - accuracy: 0.8230
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.3747 - accuracy: 0.8658
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.3363 - accuracy: 0.8773
Epoch 4/5
1875/1875 [==============================] - 5s 2ms/step - loss: 0.3094 - accuracy: 0.8856
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2926 - accuracy: 0.8901


The result tell us that after 5 epochs the accuracy of the model is 89%, meaning that it classified 89% of the training set correctly.

## Test the model

Now we feed the test dataset to check how well the NN predicts data that has not seen.

In [10]:
model.evaluate(test_images, test_labels)

313/313 [==============================] - 1s 2ms/step - loss: 0.3565 - accuracy: 0.8732


[0.3565315008163452, 0.873199999332428]

In this case, the NN classified correctly 87% of the images. Not great, and by far not the best accuracy, but this is a very small example with a "basic" NN model.

Finally, in the training section you can include *callbacks* (a user define function) in model.fit to stop the training process when you reach the desire metric. For example, if your goal is 95% of accuracy and you have 1,000 epochs, you can set the training the process to stop once it reaches 95%, even if the 1,000 epochs have not finished.